&[T] can be converted into [T; N] but &mut [T] can't

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.
-->

I tried this code:

```rust
fn main() {
    use std::convert::TryFrom;

    let s: &[i32] = &[1, 2, 3, 4, 5];
    let _: [i32; 5] = TryFrom::try_from(s).unwrap();

    let s: &mut [i32] = &mut [1, 2, 3, 4, 5];
    let _: [i32; 5] = TryFrom::try_from(s).unwrap();
}
```

I expected to see this happen: The source is compiled without any error.

Instead, this happened: Compile error as follows.

```
error[E0277]: the trait bound `[i32; 5]: From<&mut [i32]>` is not satisfied
   --> foo.rs:8:23
    |
8   |     let _: [i32; 5] = TryFrom::try_from(s).unwrap();
    |                       ^^^^^^^^^^^^^^^^^ the trait `From<&mut [i32]>` is not implemented for `[i32; 5]`
    |
    = note: required because of the requirements on the impl of `Into<[i32; 5]>` for `&mut [i32]`
    = note: required because of the requirements on the impl of `TryFrom<&mut [i32]>` for `[i32; 5]`
note: required by `try_from`
   --> /Users/rhysd/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/convert/mod.rs:477:5
    |
477 |     fn try_from(value: T) -> Result<Self, Self::Error>;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `[i32; 5]: From<&mut [i32]>` is not satisfied
 --> foo.rs:8:23
  |
8 |     let _: [i32; 5] = TryFrom::try_from(s).unwrap();
  |                       ^^^^^^^^^^^^^^^^^^^^ the trait `From<&mut [i32]>` is not implemented for `[i32; 5]`
  |
  = note: required because of the requirements on the impl of `Into<[i32; 5]>` for `&mut [i32]`
  = note: required because of the requirements on the impl of `TryFrom<&mut [i32]>` for `[i32; 5]`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0277`.
```

Is there any reason `TryFrom<&mut '_ [T]> for [T; N]` is not implemented though [`TryFrom<&'_ [T]> for [T; N]` is implemented](https://doc.rust-lang.org/std/primitive.array.html#impl-TryFrom%3C%26%27_%20%5BT%5D%3E)?

Since the implementation is lack, I needed to convert the mutable slice into immutable slice like below.

```rust
let s: &mut [T] = ...;
let a: [T; N] = (&*s).try_into().unwrap();
```

### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.56.1 (59eed8a2a 2021-11-01)
binary: rustc
commit-hash: 59eed8a2aac0230a8b53e89d4e99d55912ba6b35
commit-date: 2021-11-01
host: x86_64-apple-darwin
release: 1.56.1
LLVM version: 13.0.0
```

--- a/library/core/src/array/mod.rs
+++ b/library/core/src/array/mod.rs
@@ -176,6 +176,18 @@ where
     }
 }
 
+#[stable(feature = "try_from_mut_slice_to_array", since = "1.57.0")]
+impl<T, const N: usize> TryFrom<&mut [T]> for [T; N]
+where
+    T: Copy,
+{
+    type Error = TryFromSliceError;
+
+    fn try_from(slice: &mut [T]) -> Result<[T; N], TryFromSliceError> {
+        <Self>::try_from(slice.as_ref())
+    }
+}
+
 #[stable(feature = "try_from", since = "1.34.0")]
 impl<'a, T, const N: usize> TryFrom<&'a [T]> for &'a [T; N] {
     type Error = TryFromSliceError;

--- a/library/core/src/array/mod.rs
+++ b/library/core/src/array/mod.rs
@@ -176,7 +176,7 @@ where
     }
 }
 
-#[stable(feature = "try_from_mut_slice_to_array", since = "1.58.0")]
+#[stable(feature = "try_from_mut_slice_to_array", since = "1.59.0")]
 impl<T, const N: usize> TryFrom<&mut [T]> for [T; N]
 where
     T: Copy,

--- a/library/core/src/array/mod.rs
+++ b/library/core/src/array/mod.rs
@@ -184,7 +184,7 @@ where
     type Error = TryFromSliceError;
 
     fn try_from(slice: &mut [T]) -> Result<[T; N], TryFromSliceError> {
-        <Self>::try_from(slice.as_ref())
+        <Self>::try_from(&*slice)
     }
 }
 

--- a/library/core/src/array/mod.rs
+++ b/library/core/src/array/mod.rs
@@ -176,7 +176,7 @@ where
     }
 }
 
-#[stable(feature = "try_from_mut_slice_to_array", since = "1.57.0")]
+#[stable(feature = "try_from_mut_slice_to_array", since = "1.58.0")]
 impl<T, const N: usize> TryFrom<&mut [T]> for [T; N]
 where
     T: Copy,

--- a/library/core/tests/array.rs
+++ b/library/core/tests/array.rs
@@ -28,11 +28,22 @@ fn array_try_from() {
         ($($N:expr)+) => {
             $({
                 type Array = [u8; $N];
-                let array: Array = [0; $N];
+                let mut array: Array = [0; $N];
                 let slice: &[u8] = &array[..];
 
                 let result = <&Array>::try_from(slice);
                 assert_eq!(&array, result.unwrap());
+
+                let result = <Array>::try_from(slice);
+                assert_eq!(&array, &result.unwrap());
+
+                let mut_slice: &mut [u8] = &mut array[..];
+                let result = <&mut Array>::try_from(mut_slice);
+                assert_eq!(&[0; $N], result.unwrap());
+
+                let mut_slice: &mut [u8] = &mut array[..];
+                let result = <Array>::try_from(mut_slice);
+                assert_eq!(&array, &result.unwrap());
             })+
         }
     }

